src/Microsoft.AspNetCore.Sockets/HttpConnectionDispatcher.cs

```diff
@@ -202,58 +227,54 @@ private Task ProcessSend(HttpContext context)
             ConnectionState state;
             if (_manager.TryGetConnection(connectionId, out state))
             {
-                // If we received an HTTP POST for the connection id and it's not an HttpChannel then fail.
-                // You can't write to a TCP channel directly from here.
-                var httpChannel = state.Connection.Channel as HttpConnection;
-
-                if (httpChannel == null)
+                if (state.Connection.Mode == ConnectionMode.Streaming)
                 {
-                    throw new InvalidOperationException("No channel");
+                    var streamingState = (StreamingConnectionState)state;
+
+                    await context.Request.Body.CopyToAsync(streamingState.Application.Output);
                 }
+                else
+                {
+                    // Collect the message and write it to the channel
+                    // TODO: Need to use some kind of pooled memory here.
+                    byte[] buffer;
+                    using (var strm = new MemoryStream())
+                    {
+                        await context.Request.Body.CopyToAsync(strm);
+                        await strm.FlushAsync();
+                        buffer = strm.ToArray();
+                    }
 
-                return context.Request.Body.CopyToAsync(httpChannel.Input);
+                    var format =
+                        string.Equals(context.Request.Query["format"], "binary", StringComparison.OrdinalIgnoreCase)
+                            ? Format.Binary
+                            : Format.Text;
+                    var message = new Message(
+                        ReadableBuffer.Create(buffer).Preserve(),
```

Comment:
I agree, but we don't have a Pipeline to alloc from, since we're dealing with Messaging transports. This leads to the desire to allow allocating PreservedBuffers directly out of the buffer pool.